substrate/frame/nomination-pools/src/lib.rs

```diff
@@ -3468,31 +3474,59 @@ impl<T: Config> Pallet<T> {
 		member_account: &T::AccountId,
 		reporter: Option<T::AccountId>,
 	) -> DispatchResult {
-		// calculate points to be slashed.
-		let member =
-			PoolMembers::<T>::get(&member_account).ok_or(Error::<T>::PoolMemberNotFound)?;
+		let member = PoolMembers::<T>::get(member_account).ok_or(Error::<T>::PoolMemberNotFound)?;
 
-		let pool_account = Pallet::<T>::generate_bonded_account(member.pool_id);
-		ensure!(T::StakeAdapter::has_pending_slash(&pool_account), Error::<T>::NothingToSlash);
-
-		let unslashed_balance = T::StakeAdapter::member_delegation_balance(&member_account);
-		let slashed_balance = member.total_balance();
-		defensive_assert!(
-			unslashed_balance >= slashed_balance,
-			"unslashed balance should always be greater or equal to the slashed"
-		);
+		let pending_slash =
+			Self::member_pending_slash(MemberAccount(member_account.clone()), member.clone())?;
 
 		// if nothing to slash, return error.
-		ensure!(unslashed_balance > slashed_balance, Error::<T>::NothingToSlash);
+		ensure!(!pending_slash.is_zero(), Error::<T>::NothingToSlash);
 
 		T::StakeAdapter::member_slash(
-			&member_account,
-			&pool_account,
-			unslashed_balance.defensive_saturating_sub(slashed_balance),
+			MemberAccount(member_account.clone()),
+			PoolAccount(Pallet::<T>::generate_bonded_account(member.pool_id)),
+			pending_slash,
 			reporter,
 		)
 	}
 
+	/// Pending slash for a member.
+	///
+	/// Takes the pool_member object corresponding to the `member_account`.
+	fn member_pending_slash(
+		member_account: MemberAccount<T::AccountId>,
+		pool_member: PoolMember<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		// only executed in tests: ensure the member account is correct.
+		debug_assert!(
+			PoolMembers::<T>::get(member_account.0.clone()).expect("member must exist") ==
+				pool_member
+		);
+
+		let pool_account = Pallet::<T>::generate_bonded_account(pool_member.pool_id);
+		// if the pool doesn't have any pending slash, it implies the member also does not have any
```

Comment:
Why do we perform this check before? It seems that checking if the pool doesn't have any pending slash is as expensive as retrieving the member delegation balance.

Thus we end up paying for an extra read if the member is to be slashed and not gaining anything in return.

Reply:
`pool_member.total_balance()` also does couple of extra reads along with iterating over the unbonding eras. But your point is fair if we look only at `DelegateStake` adapter. 

My intention was to short circuit these calls without any extra reads for `TransferStake` adapter (its impl just returns zero). Once we remove them (if we ever remove them), we probably can remove this check.